### PR TITLE
CBG-1365: Calculate channels to revoke

### PR DESCRIPTION
**Pull request by JRascagneres**
Uses the built history to calculate channels to revoke along with the triggeredBy value for each of the channels.

### CODE REVIEW

File: auth/user.go [hunk 1]
```diff
@@ -183,6 +183,98 @@ func (user *userImpl) RoleHistory() TimedSetHistory {
 	return user.RoleHistory_
 }
 
+func (user *userImpl) RevokedChannels(since uint64) map[string]uint64 {
```
Comment (adamcfraser) [inline auth/user.go]: Add a comment describing the contents of the returned map[string]uint64.  Even after reviewing the entire function I'm not positive what the sequence value represents (it seems like 'end seq for latest revocation', but I'm not sure how that information should be used).

File: auth/user.go [hunk 2]
```diff
@@ -183,6 +183,98 @@ func (user *userImpl) RoleHistory() TimedSetHistory {
...
+
+	// Grab all channels user has access to
```
Comment (adamcfraser) [inline auth/user.go]: This comment is probably unnecessary - the call to InheritedChannels() is self-documenting.

File: auth/user.go [hunk 3]
```diff
@@ -183,6 +183,98 @@ func (user *userImpl) RoleHistory() TimedSetHistory {
...
+	accessibleChannels := user.InheritedChannels()
+
+	// Get revoked roles
+	rolesToRevoke := map[string]uint64{}
+	roleHistory := user.RoleHistory()
+	for roleName, history := range roleHistory {
+		if !user.RoleNames().Contains(roleName) {
+			for _, entry := range history.Entries {
+				if entry.StartSeq <= since && entry.EndSeq >= since {
```
Comment (adamcfraser) [inline auth/user.go]: Having both StartSeq and EndSeq comparisons as "or equal to" doesn't look correct - my guess is that it should be simply `entry.EndSeq > since`, as a call with since=endSeq usually implies everything up to and including endSeq has already been replicated.  But unit tests with boundary conditions would be the best approach.

File: auth/user.go [hunk 4]
```diff
@@ -183,6 +183,98 @@ func (user *userImpl) RoleHistory() TimedSetHistory {
...
+					mostRecentEndSeq := history.Entries[len(history.Entries)-1]
+					rolesToRevoke[roleName] = mostRecentEndSeq.EndSeq
+				}
+			}
+		}
+	}
+
+	// Store revoked channels to return
```
Comment (adamcfraser) [inline auth/user.go:222]: Are both of these inline functions strictly necessary?  For addToCombined, it seems like this would be cleaner as a custom type for the map[string]uint64, with a function like "add(string, uint64)" (that did the check addToCombined is doing).

For revokeChannelHistoryProcessing, what is the benefit of using as an inline function instead of a standalone function?

**Review by adamcfraser — CHANGES_REQUESTED** [review]
A few comments, and  I'd like a walkthrough of RevokedChannels once those have been dealt with, have a few questions that are better talked through.

File: auth/user.go (hunk 1 above)
Comment (JRascagneres) [reply auth/user.go]: Basically the value that should be used for triggeredBy, but I'll add a comment yeah.

File: auth/user.go (hunk 4 above)
Comment (JRascagneres) [reply auth/user.go:222]: I like the idea for the custom type, that works.
revokeChannelHistoryProcessing - benefit was just its only used internally for that function so seems unnecessary to have it 'bloat' externally but I can do. When a function only ever needs to be used inside another I tend to go for these inline functions. Also has access to the scope in this case to add to the map directly rather than via a parameter.

File: auth/user.go (hunk 3 above)
Comment (JRascagneres) [reply auth/user.go]: Yeah, you're correct would end up with unnecessary revocation message. Confirmed with test.

File: auth/user.go (hunk 4 above)
Comment (JRascagneres) [reply auth/user.go:222]: Note to myself: Addressed first part, second part not yet addressed. Can maybe discuss on Monday along with other.

**Review by adamcfraser — APPROVED** [review]
